govcd/vm.go

```diff
@@ -164,53 +205,80 @@ func (vm *VM) ChangeCPUCountWithCore(virtualCpuCount int, coresPerSocket *int) (
 
 }
 
-func (vm *VM) ChangeNetworkConfig(networks []map[string]interface{}, ip string) (Task, error) {
-	err := vm.Refresh()
-	if err != nil {
-		return Task{}, fmt.Errorf("error refreshing VM before running customization: %v", err)
-	}
+func (vm *VM) updateNicParameters(networks []map[string]interface{}, networkSection *types.NetworkConnectionSection) {
 
-	networkSection, err := vm.GetNetworkConnectionSection()
+	for tfNicSlot, network := range networks {
+		for loopIndex := range networkSection.NetworkConnection {
+			// Change network config only if we have the same virtual slot number as in .tf config
+			if tfNicSlot == networkSection.NetworkConnection[loopIndex].NetworkConnectionIndex {
 
-	// changes network config when only matches network name with provided one
-	for _, network := range networks {
-		for index, networkConnection := range networkSection.NetworkConnection {
-			if networkConnection.Network == network["orgnetwork"] { // network name are equal
 				// Determine what type of address is requested for the vApp
 				ipAllocationMode := types.IPAllocationModeNone
 				ipAddress := "Any"
 
-				// TODO: Review current behaviour of using DHCP when left blank
-				if ip == "dhcp" || network["ip"].(string) == "dhcp" {
+				var ipFieldString string
+
+				ipField, ipIsSet := network["ip"]
+				if ipIsSet {
+					ipFieldString = ipField.(string)
+				}
+
+				switch {
+				// TODO 3.0 remove from here when legacy `ip` and `network_name` attributes are removed
+				case ipIsSet && ipFieldString == "dhcp": // Legacy ip="dhcp" mode
 					ipAllocationMode = types.IPAllocationModeDHCP
-				} else if ip == "allocated" || network["ip"].(string) == "allocated" {
+				case ipIsSet && ipFieldString == "allocated": // Legacy ip="allocated" mode
 					ipAllocationMode = types.IPAllocationModePool
-				} else if ip == "none" || network["ip"].(string) == "none" {
+				case ipIsSet && ipFieldString == "none": // Legacy ip="none" mode
 					ipAllocationMode = types.IPAllocationModeNone
-				} else if ip != "" {
-					ipAllocationMode = types.IPAllocationModeManual
-					// TODO: Check a valid IP has been given
-					ipAddress = ip
-				} else if network["ip"].(string) != "" {
+
+				// Legacy ip="valid_ip" mode (currently it is hit by ip_allocation_mode=MANUAL as well)
```

Comment:
I suggest chaning word "legacy" to "deprecated" in all comments to be more precise.

Reply:
Addressed